changes.go

```diff
@@ -22,54 +22,294 @@ func FromData(data *charm.BundleData) []*Change {
+func newAddMachineChange(args AddMachineArgs, requires ...string) *AddMachineChange {
+	return &AddMachineChange{
+		changeInfo: changeInfo{
+			requires: requires,
+			method:   "addMachines",
+		},
+		Args: args,
+	}
+}
+
+// AddMachineChange holds a change for adding a machine or container.
+type AddMachineChange struct {
+	changeInfo
+	// Args holds parameters for adding a machine.
+	Args AddMachineArgs
+}
+
+// GUIArgs implements Change.GUIArgs.
+func (ch *AddMachineChange) GUIArgs() []interface{} {
+	options := AddMachineOptions{
+		Series:        ch.Args.Series,
+		Constraints:   ch.Args.Constraints,
+		ContainerType: ch.Args.ContainerType,
+		ParentId:      ch.Args.ParentId,
+	}
+	return []interface{}{options}
+}
+
+// AddMachineOptions holds GUI options for adding a machine or container.
+type AddMachineOptions struct {
+	// Series holds the machine OS series.
+	Series string `json:"series,omitempty"`
+	// Constraints holds the machine constraints.
+	Constraints string `json:"constraints,omitempty"`
+	// ContainerType holds the machine container type (like "lxc" or "kvm").
+	ContainerType string `json:"containerType,omitempty"`
+	// ParentId holds the id of the parent machine.
+	ParentId string `json:"parentId,omitempty"`
+}
+
+// AddMachineArgs holds parameters for adding a machine or container.
+type AddMachineArgs struct {
+	// Series holds the optional machine OS series.
+	Series string
+	// Constraints holds the optional machine constraints.
+	Constraints string
+	// ContainerType optionally holds the type of the container (for instance
+	// ""lxc" or kvm"). It is not specified for top level machines.
+	ContainerType string
+	// ParentId optionally holds a placeholder pointing to another machine
+	// change or to a unit change. This value is only specified in the case
+	// this machine is a container, in which case also ContainerType is set.
+	ParentId string
+}
+
+// newAddRelationChange creates a new change for adding a relation.
+func newAddRelationChange(args AddRelationArgs, requires ...string) *AddRelationChange {
+	return &AddRelationChange{
+		changeInfo: changeInfo{
+			requires: requires,
+			method:   "addRelation",
+		},
+		Args: args,
+	}
 }
 
-// add is an addChangeFunc that can be used to add a change to this change set.
-func (cs *changeset) add(method string, requires []string, args ...interface{}) *Change {
-	if args == nil {
-		args = make([]interface{}, 0)
+// AddRelationChange holds a change for adding a relation between two services.
+type AddRelationChange struct {
+	changeInfo
+	// Args holds parameters for adding a relation.
+	Args AddRelationArgs
+}
+
+// GUIArgs implements Change.GUIArgs.
+func (ch *AddRelationChange) GUIArgs() []interface{} {
+	return []interface{}{ch.Args.Endpoint1, ch.Args.Endpoint2}
+}
+
+// AddRelationArgs holds parameters for adding a relation between two services.
+type AddRelationArgs struct {
+	// Endpoint1 and Endpoint2 hold relation endpoints, like "$deploy-1:web" or
```

Comment:
What's the "-1" syntax here? I'd prefer to see an explicit description of the possible syntax with the strings as examples.